.github/workflows/main.yml

```diff
@@ -19,25 +19,77 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
-      # TODO: cache all the setup
-
       - name: Setup directories
         run: |
           # create directory that will be on the PATH
           mkdir -p ~/.local/bin
           echo "${HOME}/.local/bin" >> $GITHUB_PATH
           mkdir -p ~/source
+     
+      - uses: actions/cache@v2
+        id: cache-gcc
+        with:
+          path: ~/.local/bin/gcc
+          key: ${{ runner.os }}-gcc
+
+      - uses: actions/cache@v2
+        id: cache-clang
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-clang
+      
+      - uses: actions/cache@v2
+        id: cache-clang-format
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-clang-format
+
+      - uses: actions/cache@v2
+        id: cache-libc6-i386
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-libc6-i386
+
+      - uses: actions/cache@v2
+        id: cache-linux-module
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-linux-module
+
+      - uses: actions/cache@v2
+        id: cache-make
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-make
```

Comment:
I'm not sure where clang, clang-format, make, etc are installed, but it's definitely not in `~/.cache/pip`! You'll need to figure out where the actual clang, make, etc binaries get downloaded and installed to and use those paths. This might involve reading the source of any actions in a `uses` step (like the one that installs `arm-none-eabi-gcc`) or running the commands on your machine and figuring out where things got installed to.

Also I notice there's no cache step for the installing arm-none-eabi-gcc step (`uses: fiam-arm-none-eabi-gcc@v1`)? It installs the toolchain we need to build for STM32, and it takes almost two full minutes to do so - we'll definitely need to cache that step.